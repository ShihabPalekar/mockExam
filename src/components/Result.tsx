const DetailsForm = () => {
    return(
        <div className="result">

        </div>
    )
}

export default DetailsForm