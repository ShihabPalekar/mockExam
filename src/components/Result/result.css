.results-wrapper{
    display: flex;
    flex-direction: row;
    margin-left: 25vw;
}

.results-wrapper > div {
    background-color: white;
    padding: 25px;
    text-align: left;
    margin: 5px 40px;
}
