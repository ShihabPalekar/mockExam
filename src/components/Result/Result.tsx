import { useEffect, useState } from "react";
import { PieChart } from "react-minimal-pie-chart";
import "./result.css";

type Props = {
  data: any;
};
const Result: React.FC<Props> = ({ ...props }) => {
  const [isLoading, setIsLoading] = useState(true);
  const { data } = props;
  const [totalMarks, setTotalMarks] = useState(0);
  const [marks, setMarks] = useState(0);

  useEffect(() => {
    for (let i = 0; i < data.length; i++) {
      if (data[i].type === "match-up") {
        let correct = data[i].ansInput
          ?.map((op: any) => op.answer === op.selectedOpt)
          .filter((i: any) => i === true).length;
        if (!isNaN(correct)) {
          setMarks((prevMarks) => prevMarks + 1);
        }
        setTotalMarks(
          (prevTotalMarks) => prevTotalMarks + data[i].columnA.length
        );
      } else if (data[i].type === "multi-select") {
        if (
          JSON.stringify(data[i].answer.sort()) ===
          JSON.stringify(data[i].ansInput.sort())
        ) {
          setMarks((prevMarks) => prevMarks + 1);
        }
        setTotalMarks((prevTotalMarks) => prevTotalMarks + 1);
      } else {
        data[i].answer === data[i].ansInput &&
          setMarks((prevMarks) => prevMarks + 1);
        setTotalMarks((prevTotalMarks) => prevTotalMarks + 1);
      }
    }
    setIsLoading(false);
  }, []);

  console.log(marks);

  if (isLoading) {
    return <p className="loader">Calculating result...</p>;
  }

  return (
    <div className="results-wrapper">
      <div className="correct-ans">
        <h2 role="ans-header">Answers</h2>
        {data.map((i: any, index: number) => {
          if (i.type === "match-up") {
            return (
              <div>
                <p>{index + 1}.</p>
                <div>
                  <p>Your answer:</p>
                  {i.ansInput.map((a: any) => (
                    <span style={{ marginRight: "10px" }}>{a.selectedOpt}</span>
                  ))}
                </div>
                <div>
                  <p>Correct answer:</p>
                  {i.ansInput.map((a: any) => (
                    <span style={{ marginRight: "10px" }}>{a.answer}</span>
                  ))}
                </div>
              </div>
            );
          } else if (i.type === "multi-select") {
            return (
              <div>
                <p>{index + 1}.</p>
                <div>
                  <span style={{ marginRight: "10px" }}>Your answer:</span>
                  {i.ansInput.map((a: string) => (
                    <span style={{ marginRight: "10px" }}>{a}</span>
                  ))}
                </div>
                <br />
                <div>
                  <span style={{ marginRight: "10px" }}>Correct answer:</span>
                  {i.answer.map((a: string) => (
                    <span style={{ marginRight: "10px" }}>{a}</span>
                  ))}
                </div>
              </div>
            );
          } else {
            return (
              <div>
                <p>{index + 1}.</p>
                <div>
                  <span style={{ marginRight: "10px" }}>Your answer:</span>
                  <span>{i.ansInput}</span>
                </div>
                <br />
                <div>
                  <span style={{ marginRight: "10px" }}>Correct answer:</span>
                  <span>{i.answer}</span>
                </div>
              </div>
            );
          }
        })}
      </div>
      <div className="pie">
        <h2>Result</h2>
        <h3 style={{marginTop: "25px"}} role="marks-scored">Your total score is {marks} out of {totalMarks}.</h3>
        <PieChart
          totalValue={totalMarks}
          data={[
            {
              title: `Correct answers ${(marks / totalMarks) * 100}`,
              value: marks,
              color: "green",
            },
            {
              title: `Incorrect answers  ${
                ((totalMarks - marks) / totalMarks) * 100
              }`,
              value: totalMarks - marks,
              color: "red",
            },
          ]}
        />
      </div>
    </div>
  );
};

export default Result;
