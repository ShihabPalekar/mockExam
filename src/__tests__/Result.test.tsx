import React from "react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { fireEvent, render, screen } from "@testing-library/react";
import Result from "../components/Result/Result";

const mockData = [
  {
    id: 5,
    type: "multi-select",
    que: "Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem.",
    options: ["dummy1", "dummy2", "dummy3", "dummy4"],
    answer: ["dummy2", "dummy4"],
    ansInput: ["dummy2", "dummy4"],
  },
];

test("snapshot of result page", () => {
  const component = renderer.create(<Result data={undefined} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Result component", () => {
  test("renders result component", () => {
    render(<Result data={mockData} />);
  });

  test("correct answers are shown", () => {
    render(<Result data={mockData} />);
    let answers = screen.getByRole("ans-header");
    expect(answers).toBeInTheDocument();
  });

  test("marks scored by user are shown", () => {
    render(<Result data={mockData} />);
    let marks = screen.getByRole("marks-scored");
    expect(marks).toBeInTheDocument();
  });

});
