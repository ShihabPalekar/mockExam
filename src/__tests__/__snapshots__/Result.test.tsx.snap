// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of result page 1`] = `
<p
  className="loader"
>
  Calculating result...
</p>
`;
